pyvips/vimage.py

```diff
@@ -125,6 +125,107 @@ def _dep(func):
+        the number of bands
+    format : str
+        the libvips format of the image
+
+    Returns
+    -------
+    interpretation : str
+        the best guess for the appropriate libvips format.
+
+
+    This function is a helper for `Image` creation from arrays when the user
+    specifies 'auto' for `interpretation`.
+
+    The heuristic used here can be defined in the following table:
+
+    | bands | format    | interpretation | reason       |
+    | ----- | --------- | -------------- | ------------ |
+    | 1     | uchar     | b-w            | L            |
+    | 2     | uchar     | b-w            | LA           |
+    | 3     | uchar     | rgb            | RGB          |
+    | 4     | uchar     | rgb            | RGBA         |
+    | 5+    | uchar     | multiband      | spectral     |
+    | 1     | char      | matrix         | conv. kernel |
+    | 2+    | char      | multiband      |              |
+    | 1     | ushort    | grey16         | L            |
+    | 2     | ushort    | grey16         | LA           |
+    | 3     | ushort    | rgb16          | RGB          |
+    | 4     | ushort    | rgb16          | RGBA         |
+    | 5+    | ushort    | multiband      | spectral     |
+    | any   | short     | multiband      |              |
+    | 1     | float     | b-w            | L            |
+    | 2     | float     | b-w            | LA           |
+    | 3     | float     | scrgb          | RGB          |
+    | 4     | float     | scrgb          | RGBA         |
+    | 5+    | float     | multiband      | spectral     |
+    | 1     | double    | b-w            | L            |
+    | 2     | double    | b-w            | LA           |
+    | 3     | double    | scrgb          | RGB          |
+    | 4     | double    | scrgb          | RGBA         |
+    | 5+    | double    | multiband      | spectral     |
+    | any   | complex   | fourier        | FFT          |
+    | any   | dpcomplex | fourier        | FFT          |
+
+    '''
+
+    if format not in FORMAT_TO_TYPESTR:
+        raise ValueError('Unknown format: {}'.format(format))
+    if not isinstance(bands, int) or bands < 1:
+        raise ValueError('Number of bands must be a positive integer.')
+
+    interp = 'multiband' # the default
+
+    if format == 'uchar':
+        if bands in [1, 2]:
+            interp = 'b-w'
+        elif bands in [3, 4]:
+            interp = 'rgb'
+    elif format == 'char':
+        if bands == 1:
+            interp = 'matrix'
+    elif format == 'ushort':
+        if bands in [1, 2]:
+            interp = 'grey16'
+        elif bands in [3, 4]:
+            interp = 'rgb16'
+    elif format == 'short':
+        if bands == 1:
+            interp = 'matrix'
+    elif format in ['float', 'double']:
+        if bands in [1, 2]:
+            interp = 'b-w'
+        elif bands in [3, 4]:
+            interp = 'scrgb'
+    elif format in ['complex', 'dpcomplex']:
+        interp = 'fourier'
+
+    return interp
```

Comment:
libvips has something like this too: 

https://www.libvips.org/API/current/libvips-header.html#vips-image-guess-interpretation

It looks at the current interpretation and tries to sanity check it. If it looks impossible, it attempts to guess a sane interpretation. If you set `interpretation` to `"error"`, you should trigger this sanitiser (I think).

Maybe we can combine the two sets of rules? You can see the one libvips uses here:

https://github.com/libvips/libvips/blob/master/libvips/iofuncs/header.c#L555-L721